File: src/main/java/com/hanghai/kchtg/beacon/entity/Buoy.java
package com.hanghai.kchtg.beacon.entity;

import com.hanghai.kchtg.common.entity.ApprovalStatus;
import com.hanghai.kchtg.common.entity.BaseEntity;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.PrePersist;
import jakarta.persistence.Table;
import jakarta.validation.constraints.*;
import lombok.*;
import org.hibernate.annotations.SQLRestriction;

import java.time.LocalDate;

/**
 * Entity representing nautical buoy equipment (cardinal, sector, special, safe water, isolated danger).
 * Extends BaseEntity for shared audit fields and soft-delete support.
 */
@Entity
@Table(name = "buoy")
@SQLRestriction("deleted_at IS NULL")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Buoy extends BaseEntity {
    @Column(name = "province_id")
    private Integer provinceId;


    @NotBlank(message = "Mã phao tiêu không được để trống")
    @Size(max = 50)
    @Column(nullable = false, unique = true, length = 50)
    private String code;

    @NotBlank(message = "Tên phao tiêu không được để trống")
    @Size(max = 255)
    @Column(nullable = false, length = 255)
    private String name;

    @Column
    private String type;


    @Size(max = 50)
    private String color;

    @Size(max = 50)
    private String shape;

    @Size(max = 100)
    @Column(name = "light_characteristic", length = 100)
    private String lightCharacteristic;

    @NotNull
    @DecimalMin("0.01")
    @DecimalMax("100.0")
    private Double range;

    @Size(max = 1000)
    private String description;

    @Column(name = "unit_id")
    private java.util.UUID unitId;

    @Column(name = "last_inspection_date")
    private LocalDate lastInspectionDate;

    @Column(name = "next_inspection_date")
    private LocalDate nextInspectionDate;

    @Column(name = "is_active")
    @Builder.Default
    private Boolean isActive = true;

    @Column(nullable = false)
    @Builder.Default
    private String status = "DRAFT";

    @Column(name = "approval_status", nullable = false)
    @Enumerated(EnumType.STRING)
    private ApprovalStatus approvalStatus;

    @Column(name = "approval_level")
    private Integer approvalLevel;

    @Column(name = "approved_by")
    private java.util.UUID approvedBy;

    @Column(name = "approved_date")
    private java.time.LocalDateTime approvedDate;

    @Column(name = "submitted_for_approval_by")
    private java.util.UUID submittedForApprovalBy;

    @Column(name = "submitted_for_approval_at")
    private java.time.LocalDateTime submittedForApprovalAt;

    @Column(name = "level1_approved_by")
    private java.util.UUID level1ApprovedBy;

    @Column(name = "level1_approved_date")
    private java.time.LocalDateTime level1ApprovedDate;

    @Column(name = "level2_approved_by")
    private java.util.UUID level2ApprovedBy;

    @Column(name = "level2_approved_date")
    private java.time.LocalDateTime level2ApprovedDate;

    @Column(name = "rejection_reason", length = 500)
    private String rejectionReason;

    @Column(name = "geometry_type", length = 20)
    private String geometryType;

    @Column(name = "map_symbol_id")
    private java.util.UUID mapSymbolId;

    @Column(name = "coordinate_system")
    private Integer coordinateSystem;

    @Column(name = "display_rule", length = 255)
    private String displayRule;

    @Column(name = "spatial_id")
    private java.util.UUID spatialId;

    // ── Các trường bổ sung theo đặc tả CSV 'QL Phao tiêu' (56 trường) ──
    @Column(name = "buoy_station_id")
    private java.util.UUID buoyStationId;

    @Size(max = 100)
    private String classification;

    @Size(max = 100)
    @Column(name = "classification_buoy")
    private String classificationBuoy;

    @Size(max = 100)
    @Column(name = "classification_mark")
    private String classificationMark;

    @Size(max = 500)
    @Column(name = "location_detail")
    private String locationDetail;

    @Size(max = 100)
    private String condition;

    @Size(max = 2000)
    private String structure;

    private Double area;

    @Column(name = "body_height")
    private Double bodyHeight;

    private Double diameter;

    @Size(max = 100)
    @Column(name = "beacon_light")
    private String beaconLight;

    @Column(name = "tower_height")
    private Double towerHeight;

    @Column(name = "light_height")
    private Double lightHeight;

    @Size(max = 100)
    @Column(name = "light_model")
    private String lightModel;

    @Size(max = 200)
    @Column(name = "tower_color")
    private String towerColor;

    @Size(max = 500)
    @Column(name = "power_supply")
    private String powerSupply;

    @Column(name = "commissioned_date")
    private LocalDate commissionedDate;

    @Column(name = "last_repair_date")
    private LocalDate lastRepairDate;

    @Size(max = 50)
    @Column(name = "light_color")
    private String lightColor;

    @Size(max = 50)
    @Column(name = "flash_type")
    private String flashType;

    @Size(max = 50)
    private String period;

    // ── Các trường bổ sung theo đặc tả CSV 'QL Phao tiêu' (STT 41/44 — Nội dung phê duyệt) ──
    @Size(max = 1000)
    @Column(name = "level1_approval_content")
    private String level1ApprovalContent;

    @Size(max = 1000)
    @Column(name = "level2_approval_content")
    private String level2ApprovalContent;

    // ── Thông tin vận hành khai thác (CSV STT 45-48, read-only) ──
    @Size(max = 100)
    @Column(name = "operation_plan_code")
    private String operationPlanCode;

    @Size(max = 255)
    @Column(name = "operation_plan_name")
    private String operationPlanName;

    @Column(name = "operation_start_date")
    private String operationStartDate;

    @Column(name = "operation_end_date")
    private String operationEndDate;

    // ── Thông tin bảo trì (CSV STT 49-52, read-only) ──
    @Size(max = 100)
    @Column(name = "maintenance_plan_code")
    private String maintenancePlanCode;

    @Size(max = 255)
    @Column(name = "maintenance_plan_name")
    private String maintenancePlanName;

    @Column(name = "maintenance_start_time")
    private String maintenanceStartTime;

    @Column(name = "maintenance_end_time")
    private String maintenanceEndTime;

    // ── Thông tin sự cố (CSV STT 53-56, read-only) ──
    @Size(max = 100)
    @Column(name = "incident_code")
    private String incidentCode;

    @Size(max = 100)
    @Column(name = "incident_type")
    private String incidentType;

    @Size(max = 500)
    @Column(name = "incident_location")
    private String incidentLocation;

    @Column(name = "incident_time")
    private String incidentTime;

    @PrePersist
    protected void onPrePersist() {
        if (approvalStatus == null) approvalStatus = ApprovalStatus.PENDING_APPROVAL;
    }
}

File: frontend/src/types/buoy.ts
// ── Buoy Types (Phao, tiêu hàng hải) ──────────────────────────────────
// Độc lập hoàn toàn với Beacon (Đèn biển)

export type BuoyType = 'CARDINAL' | 'SECTOR' | 'SPECIAL' | 'SAFE_WATER' | 'ISOLATED_DANGER';

export const BUOY_TYPE_OPTIONS: { value: BuoyType; label: string }[] = [
  { value: 'CARDINAL', label: 'Phao giới hạn hai bên' },
  { value: 'SECTOR', label: 'Phao phân luồng' },
  { value: 'SPECIAL', label: 'Phao chuyên dùng' },
  { value: 'SAFE_WATER', label: 'Phao vùng nước an toàn' },
  { value: 'ISOLATED_DANGER', label: 'Phao chướng ngại vật cô lập' },
];

export const BUOY_TYPE_MAP: Record<BuoyType, { color: string }> = {
  CARDINAL: { color: 'orange' },
  SECTOR: { color: 'blue' },
  SPECIAL: { color: 'purple' },
  SAFE_WATER: { color: 'green' },
  ISOLATED_DANGER: { color: 'red' },
};

// ── Buoy Status ─────────────────────────────────────────────────────

export type BuoyStatus =
  | 'DRAFT'
  | 'PENDING_APPROVAL'
  | 'APPROVED_LEVEL1'
  | 'REJECTED_LEVEL1'
  | 'REJECTED_LEVEL2'
  | 'APPROVED'
  | 'PROPOSED'
  | 'APPROVED_LEVEL2'
  | 'REJECTED'
  | 'ARCHIVED'
  | string;

export const BUOY_STATUS_MAP: Record<string, { color: string; label: string }> = {
  DRAFT: { color: 'default', label: 'Lưu tạm' },
  PROPOSED: { color: 'orange', label: 'Chờ phê duyệt cấp Cảng vụ/Chi cục' },
  PENDING: { color: 'orange', label: 'Chờ phê duyệt cấp Cảng vụ/Chi cục' },
  PENDING_APPROVAL: { color: 'orange', label: 'Chờ phê duyệt cấp Cảng vụ/Chi cục' },
  APPROVED_L1: { color: 'blue', label: 'Chờ phê duyệt cấp Cục' },
  APPROVED_LEVEL1: { color: 'blue', label: 'Chờ phê duyệt cấp Cục' },
  APPROVED_L2: { color: 'green', label: 'Đã phê duyệt' },
  APPROVED_LEVEL2: { color: 'green', label: 'Đã phê duyệt' },
  PUBLISHED: { color: 'green', label: 'Đã phê duyệt' },
  APPROVED: { color: 'green', label: 'Đã phê duyệt' },
  REJECTED: { color: 'red', label: 'Từ chối cấp Cảng vụ/Chi cục' },
  REJECTED_L1: { color: 'red', label: 'Từ chối cấp Cảng vụ/Chi cục' },
  REJECTED_LEVEL1: { color: 'red', label: 'Từ chối cấp Cảng vụ/Chi cục' },
  REJECTED_L2: { color: 'red', label: 'Từ chối cấp Cục' },
  REJECTED_LEVEL2: { color: 'red', label: 'Từ chối cấp Cục' },
  ARCHIVED: { color: 'default', label: 'Đã xóa' },
};

// ── Entity Interfaces ───────────────────────────────────────────────

export interface Buoy {
  [key: string]: any;
  id: string;
  name: string;
  code: string;
  type: BuoyType;
  latitude: number;
  longitude: number;
  geometryType?: string;
  mapSymbolId?: string;
  coordinateSystem?: number;
  displayRule?: string;
  coordinates?: string;
  buoyStationId?: string;
  buoyStationName?: string;
  navigationChannelId?: string;
  navigationChannelName?: string;
  orgUnitId?: string;
  classification?: string;
  classificationBuoy?: string;
  classificationMark?: string;
  provinceId?: number;
  locationDetail?: string;
  condition?: string;
  structure?: string;
  area?: number;
  bodyHeight?: number;
  diameter?: number;
  beaconLight?: string; // Tên trường DB giữ nguyên là beacon_light, ngữ nghĩa UI: 'Đèn hiệu'
  towerHeight?: number;
  lightHeight?: number;
  lightModel?: string;
  towerColor?: string;
  powerSupply?: string;
  commissionedDate?: string;
  lastRepairDate?: string;
  lightColor?: string;
  flashType?: string;
  period?: string;
  level1ApprovalContent?: string;
  level2ApprovalContent?: string;
  operationPlanCode?: string;
  operationPlanName?: string;
  operationStartDate?: string;
  operationEndDate?: string;
  maintenancePlanCode?: string;
  maintenancePlanName?: string;
  maintenanceStartTime?: string;
  maintenanceEndTime?: string;
  incidentCode?: string;
  incidentType?: string;
  incidentLocation?: string;
  incidentTime?: string;
  color?: string;
  shape?: string;
  lightCharacteristic?: string;
  range: number;
  description?: string;
  unitId?: string;
  unitName?: string;
  lastInspectionDate?: string;
  nextInspectionDate?: string;
  isActive: boolean;
  status: BuoyStatus;
  approvalStatus: string;
  approvalLevel?: number;
  approvedBy?: number;
  approvedDate?: string;
  level1ApprovedBy?: number;
  level1ApprovedDate?: string;
  level2ApprovedBy?: number;
  level2ApprovedDate?: string;
  rejectionReason?: string;
  createdBy?: number;
  updatedBy?: number;
  createdAt: string;
  updatedAt: string;
}

export interface CreateBuoyRequest {
  name: string;
  code: string;
  type?: BuoyType;
  latitude: number;
  longitude: number;
  geometryType?: string;
  mapSymbolId?: string;
  coordinateSystem?: number;
  displayRule?: string;
  coordinates?: string;
  color?: string;
  shape?: string;
  lightCharacteristic?: string;
  range: number;
  description?: string;
  unitId?: string;
  lastInspectionDate?: string;
  nextInspectionDate?: string;
  isActive?: boolean;
  action?: 'draft' | 'submit';
  buoyStationId: string;
  navigationChannelId?: string;
  orgUnitId?: string;
  classification: string;
  classificationBuoy?: string;
  classificationMark?: string;
  provinceId?: number;
  locationDetail?: string;
  condition: string;
  structure?: string;
  area?: number;
  bodyHeight?: number;
  diameter?: number;
  beaconLight?: string; // Tên trường DB giữ nguyên là beacon_light, ngữ nghĩa UI: 'Đèn hiệu'
  towerHeight?: number;
  lightHeight: number;
  lightModel?: string;
  towerColor?: string;
  powerSupply?: string;
  commissionedDate?: string;
  lastRepairDate?: string;
  lightColor?: string;
  flashType?: string;
  period?: string;
}

export interface UpdateBuoyRequest {
  name?: string;
  type?: BuoyType;
  latitude?: number;
  longitude?: number;
  geometryType?: string;
  mapSymbolId?: string;
  coordinateSystem?: number;
  displayRule?: string;
  coordinates?: string;
  color?: string;
  shape?: string;
  lightCharacteristic?: string;
  range?: number;
  description?: string;
  unitId?: string;
  orgUnitId?: string;
  lastInspectionDate?: string;
  nextInspectionDate?: string;
  isActive?: boolean;
  buoyStationId?: string;
  navigationChannelId?: string;
  classification?: string;
  classificationBuoy?: string;
  classificationMark?: string;
  provinceId?: number;
  locationDetail?: string;
  condition?: string;
  structure?: string;
  area?: number;
  bodyHeight?: number;
  diameter?: number;
  beaconLight?: string; // Tên trường DB giữ nguyên là beacon_light, ngữ nghĩa UI: 'Đèn hiệu'
  towerHeight?: number;
  lightHeight?: number;
  lightModel?: string;
  towerColor?: string;
  powerSupply?: string;
  commissionedDate?: string;
  lastRepairDate?: string;
  lightColor?: string;
  flashType?: string;
  period?: string;
}
